### PR TITLE
CAN driver

### PR DESCRIPTION
**Pull request by haitomatic**
## Summary
- CAN driver for interfacing with FPGA CAN IP Block supporting both classical CAN and CAN-FD.
- (compared to previously closed PR) Rewrite bit timing calculation function that was previously based on https://gitlab.fel.cvut.cz/canbus/ctucanfd_ip_core/-/blob/master/driver/ctucanfd_linux_defs.c#L129 which has incompatible GPL-2.0-or-later license.
- (compared to previously closed PR) move RX logic to separate work queue function instead of directly in ISR in order to avoid violating the `sem_trywait.c` assertion (https://github.com/tiiuae/nuttx/pull/68/files/ee725cce15e310172d9ec64a22ecb43c48af1346#diff-688a0321e8f972cc7e6234665177563cf41438c517cc156765aee9b36c82d9f8)
- (compared to previously closed PR) Various recommendations have also been applied.
## Impact
- Networking / SocketCAN
## Testing
- Tested with PX4 and Saluki v1,v2
- ESC control flight test result: https://ssrc.atlassian.net/wiki/spaces/~626f763fa1de4100684ec2ec/pages/559939795/Nuttx+CAN+FD+driver#CAN-DRONE-TEST-w-UAVCAN-v0

### CODE REVIEW

File: arch/risc-v/src/mpfs/mpfs_fpga_canfd.c [hunk 1]
```diff
@@ -0,0 +1,2917 @@
...
+
+      NETDEV_RXERRORS(&priv->dev);
+
+      /* Clear Data Overrun interrupt */
+
+      putreg32(MPFS_CANFD_INT_STAT_DOI,
+               priv->base + MPFS_CANFD_INT_STAT_OFFSET);
+    }
+
+  /* Clear and reset all interrupt. */
```
Comment (haitomatic) [inline arch/risc-v/src/mpfs/mpfs_fpga_canfd.c:1234]: > I don't quite understand why the interrupt routine has to manually disable the interrupt source, will this not result in missed interrupts ? Where does the source get re-enabled ?

-> this only happens in rare error scenario when there is stuck interrupt which still remains after the upper limit of the irq loop number (10000) is exceeded. In such case, we reset all interrupt and start afesh.

File: arch/risc-v/src/mpfs/mpfs_fpga_canfd.c [hunk 2]
```diff
@@ -0,0 +1,2917 @@
...
+
+  canwarn("Reset all interrupts...\n");
+  imask = 0xffffffff;
+  putreg32(imask, priv->base + MPFS_CANFD_INT_ENA_CLR_OFFSET);
+  putreg32(imask, priv->base + MPFS_CANFD_INT_ENA_SET_OFFSET);
```
Comment (haitomatic) [inline arch/risc-v/src/mpfs/mpfs_fpga_canfd.c:1239]: > I don't quite understand why the interrupt routine has to manually disable the interrupt source, will this not result in missed interrupts ? Where does the source get re-enabled ?

-> this only happens in rare error scenario when there is stuck interrupt which still remains after the upper limit of the irq loop number (10000) is exceeded. In such case, we reset all interrupt and start afesh.

File: arch/risc-v/src/mpfs/mpfs_fpga_canfd.c [hunk 3]
```diff
@@ -0,0 +1,2917 @@
...
+    }
+
+  /* Timestamp - Read and throw away */
+
+  getreg32(priv->base + MPFS_CANFD_RX_DATA_OFFSET);
+  getreg32(priv->base + MPFS_CANFD_RX_DATA_OFFSET);
+
+  /* Data */
+
+  for (i = 0; i < len; i += 4)
```
Comment (haitomatic) [inline arch/risc-v/src/mpfs/mpfs_fpga_canfd.c:605]: > How does this work, you read the same address (memory mapped register?) over and over. Does reading it somehow automatically forward the FIFO also?

-> yes. by raeding RX buffer, the RX buffer read pointer is automatically increased if RX buffer is not empty

File: arch/risc-v/src/mpfs/mpfs_fpga_canfd.c [hunk 4]
```diff
@@ -0,0 +1,2917 @@
...
+static void mpfs_txdone_work(void *arg)
+{
+  struct mpfs_driver_s *priv = (struct mpfs_driver_s *)arg;
+
+  /* There should be space for a new TX in any event.  Poll the network for
+   * new XMIT data
+   */
+
+  net_lock();
+  devif_poll(&priv->dev, mpfs_txpoll);
```
Comment (haitomatic) [inline arch/risc-v/src/mpfs/mpfs_fpga_canfd.c:923]: > Is this really how the net <-> driver communication is supposed to work? The driver polls the interface if there is more work to do?

-> yes. Once an txt buffer is processed (packet sent), driver can poll the network for new TX if there is any. This is one of two ways new TX can be initiated. The other way is via direct callback via `priv->dev.d_txavail = mpfs_txavail`

File: arch/risc-v/src/mpfs/mpfs_fpga_canfd.c [hunk 5]
```diff
@@ -0,0 +1,2917 @@
...
+            break;
+          }
+        priv->can.bittiming.sample_point =
+          req->arbi_samplep * 10; /* In one-tenth of a percent */
+
+        if (req->data_bitrate > 4000 ||
+            req->data_bitrate < req->arbi_bitrate)
+          {
+            canerr("Data bitrate higher than 4Mbps is yet to be supported. "
+                   "Data br cannot be smaller than arbitration br.");
```
[the remark this answers is not in the document]
Comment (pussuw) [reply arch/risc-v/src/mpfs/mpfs_fpga_canfd.c:2603]: Ok if it is specified so then it is fine. Like I said, having better SNR for arbitration makes sense, as then other devices on the bus will know if it is their turn to talk. But I was not aware it was strictly _required_

File: arch/risc-v/src/mpfs/mpfs_fpga_canfd.c [hunk 6]
```diff
@@ -0,0 +1,2917 @@
...
+};
+
+enum mpfs_can_txb_command
+{
+  TXT_CMD_SET_EMPTY   = 0x01,
+  TXT_CMD_SET_READY   = 0x02,
+  TXT_CMD_SET_ABORT   = 0x04
+};
+
+#define candebug _none
```
Comment (pussuw) [inline arch/risc-v/src/mpfs/mpfs_fpga_canfd.c:157]: What's this?

File: arch/risc-v/src/mpfs/mpfs_fpga_canfd.c [hunk 7]
```diff
@@ -0,0 +1,2917 @@
...
+
+/****************************************************************************
+ * Utility definitions
+ ****************************************************************************/
+
+#ifndef min
+#define min(a, b)  (a) < (b) ? a : b
```
Comment (pussuw) [inline arch/risc-v/src/mpfs/mpfs_fpga_canfd.c:164]: I still suggest adding parentheses around a and b after the ternary operation:

((a) < (b) ? (a) : (b))

File: arch/risc-v/src/mpfs/mpfs_fpga_canfd.c [hunk 8]
```diff
@@ -0,0 +1,2917 @@
...
+   * stuck in INT_STAT reg.
+   */
+
+  canerr("Stuck interrupt (isr=%08x)\n", isr);
+
+  /* Check if any of the stuck one belongs to txb status. */
+
+  if (isr & MPFS_CANFD_INT_STAT_TXBHCI)
+    {
+#if 0
```
Comment (pussuw) [inline arch/risc-v/src/mpfs/mpfs_fpga_canfd.c:1201]: Use the flag that enables candebug to enable this snippet.

**Comment by pussuw** [comment]
Thanks for the clarifying answers!

Some nitpicking comments still but otherwise this looks good to me!

File: arch/risc-v/src/mpfs/mpfs_fpga_canfd.c (hunk 6 above)
Comment (haitomatic) [reply arch/risc-v/src/mpfs/mpfs_fpga_canfd.c]: yeah, that shouldnt be here. removed it and use caninfo instead

File: arch/risc-v/src/mpfs/mpfs_fpga_canfd.c (hunk 7 above)
Comment (haitomatic) [reply arch/risc-v/src/mpfs/mpfs_fpga_canfd.c]: yes, fixed

**Review by pussuw — APPROVED** [review]
LGTM

**Pull request merged**